### ISSUE TITLE
Zygote gradient: ERROR: Enzyme cannot deduce type 

### ISSUE THREAD

**Issue by stephans3** [SciML/SciMLSensitivity.jl#744]
Hi all,
I get an error when I want to calculate the gradient of my loss function using Zygote.

Error message (short):
```julia
warning: didn't implement memmove, using memcpy as fallback which can result in errors
warning: didn't implement memmove, using memcpy as fallback which can result in errors
ERROR: Enzyme cannot deduce type 
```

Please find complete message below.

I want to find 2 parameters for my heat equation. My problem consists of 101 equations (stiff system) and 2 parameters. 
I guess it is somehow related to the stiffness of my system and so probably refers to issue #708?

My source code is:
```julia
# Define the constants for the PDE
const λ = 50.0      # Thermal conductivity
const α = 1e-5;     # Diffusivity constant

# Discretization  
const Nx = 101;     # Number of elements x-direction
const L  = 0.1;     # Length of plate
const Δx = L/(Nx-1) # Spatial sampling
const Tf = 6000.0;  # Final time
const dt = 2.0;     # Time step width

function input_signal(time,params)
    t = time
    a = params[1];
    b = params[2];
    q = 1e-6;
    c = b*(-log(q))^(1/2);

    return  exp(a- (c * (t - 1/b))^2)
end


# Diffusion: x-direction
function diffusion_x!(dx,x,Nx, Ny, Δx) # in-place
    
    for iy in 1 : Ny
        for ix in 2 : Nx-1
            i = (iy-1)*Nx + ix
            dx[i] =  (x[i-1] - 2*x[i] + x[i+1])/Δx^2
        end
        i1 = (iy-1)*Nx + 1      # West
        i2 = (iy-1)*Nx + Nx     # East
        dx[i1] = (-2*x[i1] + 2*x[i1+1])/Δx^2
        dx[i2] = (2*x[i2-1] - 2*x[i2])/Δx^2
    end

    nothing 
end


# 1D heat equation
function heat_eq!(dx,x,p,t)       
    time = t/Tf;
    u = input_signal(time, p)
    
    diffusion_x!(dx,x,Nx,1,Δx)
  
    dx .= α * dx
    dx[1] = dx[1] + α/(λ * Δx) * u
end


# Simulation without optimization
using OrdinaryDiffEq

x0 = 300 * ones(Nx) # Intial values
tspan = (0.0, Tf)   # Time span
pinit = [15.0,10.0] # Initial parameters
alg = KenCarp4()    # Numerical integrator

prob = ODEProblem(heat_eq!,x0,tspan,pinit)
sol = solve(prob,alg, saveat = dt)

using Plots
plot(sol, legend=false)


# Control strategy
const θref = 500; # Reference temperature
function myloss(p)
    sol_pred = OrdinaryDiffEq.solve(prob, alg, p=p, saveat = dt)

    if sol_pred.retcode == :Success
        err = θref .- sol_pred[:,end]  # Temperature at final time = sol_pred[:,end]
        loss = log(sum(abs2, err)+1)
    else
        loss = Inf
    end
    return loss
end

callback = function (p, l)
    display(l)
    return false
end

using SciMLSensitivity
import ForwardDiff, Zygote
ForwardDiff.gradient(myloss, pinit) # This works 
Zygote.gradient(myloss, pinit)  # This crashes
```

Complete error message:
[error_zygote.txt](https://github.com/SciML/SciMLSensitivity.jl/files/9819763/error_zygote.txt)

As a side effect this occurs also when I want to optimize my system with `Optimization.AutoZygote()`. 

My environment:
```julia
Status `~/.julia/environments/v1.8/Project.toml`
  [a6e380b2] ControlSystems v1.5.0
  [aae7a2af] DiffEqFlux v1.52.0
  [0c46a032] DifferentialEquations v7.5.0
  [e30172f5] Documenter v0.27.23
  [7da242da] Enzyme v0.10.12
  [f6369f11] ForwardDiff v0.10.32
  [a75be94c] GalacticOptim v3.4.0
  [916415d5] Images v0.25.2
  [872c559c] NNlib v0.8.9
  [429524aa] Optim v1.7.3
  [3bd65402] Optimisers v0.2.10
  [7f7a1694] Optimization v3.9.2
  [1dea7af3] OrdinaryDiffEq v6.29.3
  [14b8a8f1] PkgTemplates v0.7.29
  [91a5bcdd] Plots v1.35.4
  [1ed8b502] SciMLSensitivity v7.10.2
  [90137ffa] StaticArrays v1.5.9
  [e88e6eb3] Zygote v0.6.49
  [10745b16] Statistics
```

**Comment by ChrisRackauckas** [comment]
I've simplified it to an example that only requires Enzyme:

```julia
using Enzyme

# Define the constants for the PDE
const λ = 50.0      # Thermal conductivity
const α = 1e-5;     # Diffusivity constant

# Discretization  
const Nx = 101;     # Number of elements x-direction
const L  = 0.1;     # Length of plate
const Δx = L/(Nx-1) # Spatial sampling
const Tf = 6000.0;  # Final time
const dt = 2.0;     # Time step width

function input_signal(time,params)
    t = time
    a = params[1];
    b = params[2];
    q = 1e-6;
    c = b*(-log(q))^(1/2);

    return  exp(a- (c * (t - 1/b))^2)
end


# Diffusion: x-direction
function diffusion_x!(dx,x,Nx, Ny, Δx) # in-place
    
    for iy in 1 : Ny
        for ix in 2 : Nx-1
            i = (iy-1)*Nx + ix
            dx[i] =  (x[i-1] - 2*x[i] + x[i+1])/Δx^2
        end
        i1 = (iy-1)*Nx + 1      # West
        i2 = (iy-1)*Nx + Nx     # East
        dx[i1] = (-2*x[i1] + 2*x[i1+1])/Δx^2
        dx[i2] = (2*x[i2-1] - 2*x[i2])/Δx^2
    end

    nothing 
end


# 1D heat equation
function heat_eq!(dx,x,p,t)       
    time = t/Tf;
    u = input_signal(time, p)
    
    diffusion_x!(dx,x,Nx,1,Δx)
  
    dx .= α * dx
    dx[1] = dx[1] + α/(λ * Δx) * u
end

x = 300 * ones(Nx) # Intial values
dx = similar(x0)
d_dx = similar(dx)
d_x = similar(x)
t = 0.0
p = [15.0,10.0]
d_p = similar(p)

heat_eq!(dx,x,p,t)

Enzyme.autodiff(heat_eq!, Duplicated(dx, d_dx), Duplicated(x, d_x), Duplicated(p, d_p), Enzyme.Const(t));
```

So closing this so it can follow up here: https://github.com/EnzymeAD/Enzyme.jl/issues/571
